toolkit/tools/internal/retry/retry.go

```diff
@@ -4,13 +4,29 @@
 package retry
 
 import (
+	"math"
 	"time"
 )
 
+func calcDelay(retryCount int, sleep time.Duration, backoffExp float64) time.Duration {
+	// Calculate an exponential backoff.
+	// The formula is: sleep * (backoffExp ^ retryCount)
+	// For example, if sleep = 1 second, backoffExp = 2.0, retryCount = 3
+	// then the delay will be 1 * (2 ^ 3) = 8 seconds.
+	expRetry := math.Pow(float64(retryCount), backoffExp)
+	return time.Duration(expRetry * float64(sleep))
+}
+
 // Run runs function up to attempts times, waiting i * sleep duration before each i-th attempt.
 func Run(function func() error, attempts int, sleep time.Duration) (err error) {
+	return RunWithExpBackoff(function, attempts, sleep, 1.0)
+}
+
+// Run runs function up to attempts times, waiting i * backoffMult * sleep duration before each i-th attempt.
+func RunWithExpBackoff(function func() error, attempts int, sleep time.Duration, backoffExp float64) (err error) {
```

Comment:
Should we add a check for `backoffExp < 0.0`?

Reply:
Technically negative numbers are not invalid exponents, although maybe not meaningful for normal use (they asymptotically approach zero, ie contant delay with each retry)

Reply:
Now treat negatives as linear backoff (ie legacy calculations for linear)